### ISSUE TITLE
SIGSEGV JVM fatal in skiko

### ISSUE THREAD

**Issue by jimgoog** [JetBrains/compose-jb#486]
Running the following code seems to cause the JVM to fatal in skiko.  Tested on Linux and Mac, was able to reproduce on both.

```
import androidx.compose.ui.platform.TestComposeWindow
import androidx.compose.ui.Modifier
import androidx.compose.material.icons.filled.AccountBox
import androidx.compose.foundation.layout.*
import androidx.compose.material.icons.Icons
import androidx.compose.material.*

fun main() {

for(i in 1..15)
    TestComposeWindow(800, 800).setContent({
        Box(Modifier.fillMaxWidth()) {
            ExtendedFloatingActionButton(
                icon = { Icon(Icons.Filled.AccountBox, "") },
                text = {},
                onClick = {},
                modifier = Modifier.fillMaxWidth()
            )
        }
    })

}
```

Error is of the form:
```
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x000000013243b39b, pid=2889, tid=6403
#
# JRE version: Java(TM) SE Runtime Environment (14.0.2+12) (build 14.0.2+12-46)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (14.0.2+12-46, mixed mode, sharing, tiered, compressed oops, g1 gc, bsd-amd64)
# Problematic frame:
# C  [libskiko-macos-x64.dylib+0xee39b]  decltype(fp((SkRecords::NoOp)())) SkRecord::Record::visit<SkRecords::Draw&>(SkRecords::Draw&) const+0xb
```

cc @igordmn

**Comment by igordmn** [comment]
Somewhere in Skija we use native object after its finalization.

Running application with "–XX:+DisableExplicitGC" don't cause crash.

It can be similar cause as in https://github.com/JetBrains/compose-jb/issues/124

**Comment by igordmn** [comment]
Or we just don't keep strong reference to skija object somewhere in Compose. Need further investigation.

**Comment by jimgoog** [comment]
I added `–XX:+DisableExplicitGC` to `jvmArgs` in the `application` block, but it still seems to crash.  I think I did see it run successfully once, so maybe it's a little non-deterministic and maybe `–XX:+DisableExplicitGC` helps a bit?  Or maybe it was just coincidence.  Unclear, but the flag certainly doesn't appear to be a viable workaround.

**Comment by olonho** [comment]
```kotlin
fun main() {

    for(i in 1..15)
        TestComposeWindow(800, 800).setContent({
            Box(Modifier.fillMaxWidth()) {
                Button(onClick = {}) {
                    Icon(Icons.Filled.AccountBox, "")
                }
            }
        })

}
```
fails as well. Guess we don't hold an icon reference in SkPicture.

**Comment by olonho** [comment]
Seems to be indeed problem with picture not being properly help while drawing into.
I.e. with logging like this:
```java
   @NotNull @Contract("_, _, _ -> this")
    public Canvas drawPicture(@NotNull Picture picture, @Nullable Matrix33 matrix, @Nullable Paint paint) {
        System.err.println("drawPictureJ " + picture.toString());
        assert picture != null : "Can’t drawPicture with picture == null";
        Stats.onNativeCall();
        _nDrawPicture(_ptr, Native.getPtr(picture), matrix == null ? null : matrix._mat, Native.getPtr(paint));
        System.err.println("drawPictureJ2 " + picture.toString());
        Reference.reachabilityFence(picture);
        Reference.reachabilityFence(paint);
        return this;
    }  
```
we get logging like this
```
Cleaning Picture 7ffa42177b70
drawPictureJ Picture(_ptr=0x7ffa42177bb0, refCount=1)
DrawPicture: 0x7ffa42177bb0
Cleaning Picture 7ffa42177bb0
drawPictureJ2 Picture(_ptr=0x7ffa42177bb0, refCount=2)
drawPictureJ Picture(_ptr=0x7ffa42178960, refCount=1)
DrawPicture: 0x7ffa42178960
drawPictureJ2 Picture(_ptr=0x7ffa42178960, refCount=1)
drawPictureJ Picture(_ptr=0x7ffa41fbc910, refCount=1)
DrawPicture: 0x7ffa41fbc910
drawPictureJ2 Picture(_ptr=0x7ffa41fbc910, refCount=2)
drawPictureJ Picture(_ptr=0x7ffa41fbc950, refCount=1)
DrawPicture: 0x7ffa41fbc950
drawPictureJ2 Picture(_ptr=0x7ffa41fbc950, refCount=2)
Cleaning Picture 7ffa41fbc910
drawPictureJ Picture(_ptr=0x7ffa41fbc990, refCount=1)
DrawPicture: 0x7ffa41fbc990
Cleaning Picture 7ffa41fbc950
Cleaning Picture 7ffa41fbc990
^^^^^^^^
#
# A fatal error has been detected by the Java Runtime Environment:
```
i.e. we clean picture `7ffa41fbc990` while it's being rendered, and neither reachability fence nor logging referring to the Picture instance prevents its cleanup.

**Comment by olonho** [comment]
Further investigation shown that this is an artifact of `GlobalSnapshotManager` implementation. Namely, when running tests, we still dispatch using `private val scheduleScope = CoroutineScope(Dispatchers.Swing + SupervisorJob())`, which leads to rendering code executed concurrently on main and UI threads, and so closing picture concurrently with rendering.
@igordmn is working on fix, which is to supply snapshot manager per window, and use proper dispatcher in real and test windows. cc: @jimgoog.

**Comment by chuckjaz** [comment]
Calling `Snapshot.sendApplyNotifications()` can happen on any thread at any time and borrows the thread it is called on to send the notifications. It is up to the receiver of the notification to ensure it handles the notification appropriately.   Having a window "local" global snapshot manager doesn't make sense to me. Can you explain more about what the proposed fix is?

**Comment by olonho** [comment]
@chuckjaz guess it's not about `Snapshot` per se, but a listener registered in `GlobalSnapshotManager`.  IMO it shall not be a global listener registered with the `Swing` dispatcher, but instead API somewhat rethought so that callbacks execution scheduling, not requesting, shall happen in controlled manner.
In a sense  `Snapshot` provides a global message bus, but the actual consumer/dispatcher for messages could be made more flexible, not just a singleton with the default delivery policy on AWT UI thread.

**Comment by chuckjaz** [comment]
I am still confused.

No code should rely on which thread or what order an apply observer is called. Calling apply observers borrows the thread of the snapshot being applied to send notifications. The snapshots are ordered but the delivery of this notification is best effort with no order guarantee other than it occurs after the snapshot has applied. 

The job of the `GlobalSnapshotManager` is just to ensure that if a write happens on the global snapshot that, eventually, a apply observer will be triggered that will have the modified object in the set of object modified. What thread that happens on just happens to be mostly the thread used by the `GlobalSnapshotManager` but might be any other thread at any time as a background thread could commit a change using `MutableSnapshot.apply()` and then the notification will occur on the thread calling `apply()`.

Any constraints, such as serialization of canvas writes, should be performed by the receiver of the message, not trying to control when and in what thread it is sent.

**Comment by olonho** [comment]
Guess it is possible that tests run headless, and so no UI thread ever exist. And overall architecture of callbacks being called on UI thread doesn’t seem to be optimal, as it will block UI while delivering callbacks, and if they could be delivered anywhere - maybe special or configurable thread would be better choice, not UI thread.

**Comment by igordmn** [comment]
Maybe it is not a big deal that `GlobalSnapshotManager` is global :).

If something capture part of the global state (like some Window/DesktopOwner) - it can know in which thread it should observe changes.

So, I think the simple fix for this issue - is to ensure that we use appropriate thread [here](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/platform/DesktopOwner.desktop.kt;l=132;drc=072a880c9ab1a245711fe9e67e9a56f2da04b790) (we can get dispatcher from [here](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/platform/DesktopOwners.desktop.kt;l=44;drc=072a880c9ab1a245711fe9e67e9a56f2da04b790))

Only one thing I don't like - we can't easily use `TestCoroutineDispatcher` for `Snapshot.sendApplyNotifications()`. But I think we can figure something out. For example, just ignore that `GlobalSnapshotManager `exists, and init some controllable `LocalSnapshotManager` in every test. We can ignore `GlobalSnapshotManager` if it doesn't produce some interfering side effects. If I am not mistaken, `TestCoroutineDispatcher` is already used in android tests, so everything should work fine.

**Comment by chuckjaz** [comment]
That is a better approach. If you need `Snapshot.sendApplyNotifications()` to occur more than the `GlobalSnapshotManager` can deliver then feel free to call it on whatever thread you need and on whatever heartbeat you need.  The`GlobalSnapshotManager` just ensures apply notifications will eventually occur, hopefully before the next frame is required by the UI. Calling it more often should not result in any correctness issues that are not already hidden bugs.

**Comment by igordmn** [comment]
Fixed in 0.4.0

**Comment by okushnikov** [comment]
Please check the <a href="https://youtrack.jetbrains.com/issue/CMP-6414">following ticket</a> on YouTrack for follow-ups to this issue. GitHub issues will be closed in the coming weeks.
